lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,127 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+	logrus "github.com/sirupsen/logrus"
+)
+
+const (
+	// PauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	PauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Generator provides a standardized way to get successive benchmarks from a consistent interface
+type Generator interface {
+	// GetBenchmark returns the benchmark config for the current generation.
+	// Returns context, Config and error
+	GetBenchmark() (context.Context, Config, error)
+	// Benchmark runs the benchmark of receiver type
+	// return an array of Results that contain information about the generations
+	Benchmark(context.Context, []string, *client.TeleportClient) ([]*Result, error)
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(ctx context.Context, lg Linear, cmd, host, login, proxy string) ([]Result, error) {
+	var result Result
+	var results []Result
+	logrus.SetLevel(logrus.ErrorLevel)
```

Comment:
logging should be managed at a higher level (by package main, usually)

Reply:
If this were to be managed at `main`, there would be an info message for each request made.

Reply:
Can't `package main` call `logrus.SetLevel` right before `benchmark.Run`?

Reply:
Yes, technically, but it still shows the info message for every request made

Reply:
This is how I am using it in `main`

```
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/gravitational/teleport/lib/benchmark"
	"github.com/logrus"
)

func main() {
	linear := &benchmark.Linear{
		LowerBound:          10,
		UpperBound:          50,
		Step:                10,
		MinimumMeasurements: 100,
		MinimumWindow:       15 * time.Second,
	}
	logrus.SetLevel(logrus.ErrorLevel)
	// Run Linear generator
	results, err := benchmark.Run(context.TODO(), linear, "ls -l /", "localhost", "quin", "localhost")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

```

Reply:
I can leave it out of `benchmark.go` but it's just a lot of extra output